### PR TITLE
selinux configs and  volume for microshift-containerized

### LINKED ISSUE #434: [BUG] setup selinux and env before starting microshift container via systemd
#### What happened:
Right now one has to manually run `CONFIG_ENV_ONLY=true ./install.sh` before starting `microshift-containerized.service`

#### What you expected to happen:
For `podman` deployment configure `selinux` policies before `systemd` calls  `ExecStart` step. 

#### Dependent on #432 and #433

### PR DESCRIPTION
**Pull request by husky-parul**
Signed-off-by: Parul Singh <parsingh@redhat.com>

**Which issue(s) this PR addresses**:
For podman deployment:

- systemd unit file for starting and managing microshift-containerized.

Closes #434, #433, #432

### CODE REVIEW

**Comment by husky-parul** [comment]
podman systemd file was created out of following command. 
```
sudo podman run --sdnotify=container --label io.containers.autoupdate=registry  --network=host --privileged -d \
            --rm --name microshift \
            -v /var/run/crio/crio.sock:/var/run/crio/crio.sock:rw,rshared \
            -v microshift-data:/var/lib/microshift:rw,rshared \
            -v /var/lib/kubelet:/var/lib/kubelet:rw,rshared \
            -v /var/log:/var/log \
            -e KUBECONFIG=/var/lib/microshift/resources/kubeadmin/kubeconfig \
            quay.io/microshift/microshift:latest
```
The volumes mounted are:
```
-v /var/run/crio/crio.sock:/var/run/crio/crio.sock:rw,rshared \
-v microshift-data:/var/lib/microshift:rw,rshared \
-v /var/lib/kubelet:/var/lib/kubelet:rw,rshared \
-v /var/log:/var/log \
```

`packaging/systemd/microshift-containerized` created `/var/lib/microshift` that gets mounted as `microshift-data` volume. 
` packaging/systemd/microshift-containerized-configure-selinux ` configure selinux policies. Both scripts are run in `ExecStartPre`

**Comment by husky-parul** [comment]
@rootfs @oglok  how are these two secrets created?` openshift-ingress/router-certs-default` and `openshift-dns/dns-default-metrics-tls`

`microshift-containerized` is behaving differently (with same selinux configurations) when mounting `host-dir ` vs `host-vol`
 
When a` host-vol  microshift-data` is used`, pods openshift-dns/dns-default-b5kwm `and` openshift-ingress/router-default-6c96f6bc66-4hnf9 `are failing because secrets `openshift-ingress/router-certs-default `and `openshift-dns/dns-default-metrics-tls  `are not present in the respective namespaces. 
(however, they are present in the name space when `host-dir` )

so 
```
sudo podman run --sdnotify=container --label io.containers.autoupdate=registry  --network=host --privileged -d \
            --rm --name microshift \
            -v /var/run/crio/crio.sock:/var/run/crio/crio.sock:rw,rshared \
            -v /var/lib/microshift:/var/lib/microshift:rw,rshared \
            -v /var/lib/kubelet:/var/lib/kubelet:rw,rshared \
            -v /var/log:/var/log \
            -e KUBECONFIG=/var/lib/microshift/resources/kubeadmin/kubeconfig \
            quay.io/microshift/microshift:latest
```
But this doesn't

```
sudo podman run --sdnotify=container --label io.containers.autoupdate=registry  --network=host --privileged -d \
            --rm --name microshift \
            -v /var/run/crio/crio.sock:/var/run/crio/crio.sock:rw,rshared \
            -v /var/lib/containers:/var/lib/containers:rw,rshared \
            -v microshift-data:/var/lib/microshift:rw,rshared \
            -v /var/lib/kubelet:/var/lib/kubelet:rw,rshared \
            -v /var/log:/var/log \
            -e KUBECONFIG=/var/lib/microshift/resources/kubeadmin/kubeconfig \
            quay.io/microshift/microshift:latest
```
The only difference is host-dir vs host-vol

**Comment by cooktheryan** [comment]
A little more information. There are some subtle differences in the mounts when doing podman inspect microshift

standard directory mount
```
"/var/lib/microshift:/var/lib/microshift:rw,rshared,rbind",
```

podman volume
```
"microshift-data:/var/lib/microshift:rw,rshared,nosuid,nodev,rbind",
```

**Comment by husky-parul** [comment]
I have narrowed down to the  serving secrets `dns-default-metrics-tls`
 and ` router-certs-default` are not created by  services `dns-default` and `router-internal-default `respectively in podman named volume deployment. Same behavior with selinux on/off. The filesystem, permissions and labels are identical for both host-dir and host-vol.

**Comment by husky-parul** [comment]
```
diff --color -Naur /tmp/config_dir.json /tmp/config_vol.json 
--- /tmp/config_dir.json        2021-11-24 13:01:39.314831482 -0500
+++ /tmp/config_vol.json        2021-11-24 13:10:36.436976929 -0500
@@ -207,21 +207,11 @@
     "oomScoreAdj": 0
   },
   "root": {
-    "path": "/var/lib/containers/storage/overlay/a9f020ec061f95a33276a901f419e9e97f0d70db8ac1be5984726f5c55393cb8/merged"
+    "path": "/var/lib/containers/storage/overlay/9e08a7c709dca5c09d54c57682dd9890e7510409e75ece238b94647b11dbdc59/merged"
   },
   "hostname": "fedora.local.com",
   "mounts": [
     {
-      "destination": "/proc",
-      "type": "proc",
-      "source": "proc",
-      "options": [
-        "nosuid",
-        "noexec",
-        "nodev"
-      ]
-    },
-    {
       "destination": "/sys",
       "type": "sysfs",
       "source": "sysfs",
@@ -234,6 +224,16 @@
       ]
     },
     {
+      "destination": "/proc",
+      "type": "proc",
+      "source": "proc",
+      "options": [
+        "nosuid",
+        "noexec",
+        "nodev"
+      ]
+    },
+    {
       "destination": "/dev",
       "type": "tmpfs",
       "source": "tmpfs",
@@ -245,28 +245,18 @@
       ]
     },
     {
-      "destination": "/run/.containerenv",
+      "destination": "/etc/resolv.conf",
       "type": "bind",
-      "source": "/run/containers/storage/overlay-containers/25ec623774d9071640cf64a913b59819f01f91ac0e13b1ce60ab1629e731a043/userdata/.containerenv",
+      "source": "/run/containers/storage/overlay-containers/09428ba251ee0e1677a9ba689265f2b6350992a219b3188b0e27ca7a8c550d08/userdata/resolv.conf",
       "options": [
         "bind",
         "rprivate"
       ]
     },
     {
-      "destination": "/var/log",
-      "type": "bind",
-      "source": "/var/log",
-      "options": [
-        "rw",
-        "rprivate",
-        "rbind"
-      ]
-    },
-    {
       "destination": "/dev/shm",
       "type": "bind",
-      "source": "/var/lib/containers/storage/overlay-containers/25ec623774d9071640cf64a913b59819f01f91ac0e13b1ce60ab1629e731a043/userdata/shm",
+      "source": "/dev/shm",
       "options": [
         "bind",
         "rprivate"
@@ -296,9 +286,19 @@
       ]
     },
     {
+      "destination": "/var/log",
+      "type": "bind",
+      "source": "/var/log",
+      "options": [
+        "rw",
+        "rprivate",
+        "rbind"
+      ]
+    },
+    {
       "destination": "/etc/hostname",
       "type": "bind",
-      "source": "/run/containers/storage/overlay-containers/25ec623774d9071640cf64a913b59819f01f91ac0e13b1ce60ab1629e731a043/userdata/hostname",
+      "source": "/run/containers/storage/overlay-containers/09428ba251ee0e1677a9ba689265f2b6350992a219b3188b0e27ca7a8c550d08/userdata/hostname",
       "options": [
         "bind",
         "rprivate"
@@ -307,25 +307,25 @@
     {
       "destination": "/etc/hosts",
       "type": "bind",
-      "source": "/run/containers/storage/overlay-containers/25ec623774d9071640cf64a913b59819f01f91ac0e13b1ce60ab1629e731a043/userdata/hosts",
+      "source": "/run/containers/storage/overlay-containers/09428ba251ee0e1677a9ba689265f2b6350992a219b3188b0e27ca7a8c550d08/userdata/hosts",
       "options": [
         "bind",
         "rprivate"
       ]
     },
     {
-      "destination": "/run/secrets",
+      "destination": "/run/.containerenv",
       "type": "bind",
-      "source": "/run/containers/storage/overlay-containers/25ec623774d9071640cf64a913b59819f01f91ac0e13b1ce60ab1629e731a043/userdata/run/secrets",
+      "source": "/run/containers/storage/overlay-containers/09428ba251ee0e1677a9ba689265f2b6350992a219b3188b0e27ca7a8c550d08/userdata/.containerenv",
       "options": [
         "bind",
         "rprivate"
       ]
     },
     {
-      "destination": "/etc/resolv.conf",
+      "destination": "/run/secrets",
       "type": "bind",
-      "source": "/run/containers/storage/overlay-containers/25ec623774d9071640cf64a913b59819f01f91ac0e13b1ce60ab1629e731a043/userdata/resolv.conf",
+      "source": "/run/containers/storage/overlay-containers/09428ba251ee0e1677a9ba689265f2b6350992a219b3188b0e27ca7a8c550d08/userdata/run/secrets",
       "options": [
         "bind",
         "rprivate"
@@ -345,19 +345,21 @@
       ]
     },
     {
-      "destination": "/var/lib/kubelet",
+      "destination": "/var/lib/microshift",
       "type": "bind",
-      "source": "/var/lib/kubelet",
+      "source": "/var/lib/containers/storage/volumes/microshift-data/_data",
       "options": [
         "rw",
         "rshared",
+        "nosuid",
+        "nodev",
         "rbind"
       ]
     },
     {
-      "destination": "/var/lib/microshift",
+      "destination": "/var/lib/kubelet",
       "type": "bind",
-      "source": "/var/lib/microshift",
+      "source": "/var/lib/kubelet",
       "options": [
         "rw",
         "rshared",
@@ -379,7 +381,7 @@
   ],
   "annotations": {
     "io.container.manager": "libpod",
-    "io.kubernetes.cri-o.Created": "2021-11-24T12:58:30.125590012-05:00",
+    "io.kubernetes.cri-o.Created": "2021-11-24T13:08:06.096394928-05:00",
     "io.kubernetes.cri-o.TTY": "false",
     "io.podman.annotations.autoremove": "TRUE",
     "io.podman.annotations.init": "FALSE",
@@ -399,15 +401,12 @@
         "limit": 2048
       }
     },
-    "cgroupsPath": "machine.slice:libpod:25ec623774d9071640cf64a913b59819f01f91ac0e13b1ce60ab1629e731a043",
+    "cgroupsPath": "machine.slice:libpod:09428ba251ee0e1677a9ba689265f2b6350992a219b3188b0e27ca7a8c550d08",
     "namespaces": [
       {
         "type": "pid"
       },
       {
-        "type": "ipc"
-      },
-      {
         "type": "uts"
       },
       {
[root@fedora ~]# 

```

File: packaging/systemd/microshift-containerized.service [hunk 1]
```diff
@@ -13,8 +13,10 @@ RequiresMountsFor=%t/containers
 Environment=PODMAN_SYSTEMD_UNIT=%n
 Restart=on-failure
 TimeoutStopSec=70
+ExecStartPre=/bin/bash /usr/bin/microshift-containerized-configure-selinux # not autogenerated
```
Comment (mangelajo) [inline packaging/systemd/microshift-containerized.service:16]:
```suggestion
ExecStartPre=/bin/bash /usr/bin/microshift-containerized-configure-selinux # not autogenerated
``` 

We should not do this, it's a lengthy process. ( installing, compiling, etc...)

Let's assume that people will download the rpms for microshift via dnf, which will include both microshift-containerized + the selinux modules compiled.

We can keep the script available just in case.
```suggestion
```

File: packaging/systemd/microshift-containerized.service [hunk 2]
```diff
@@ -13,8 +13,10 @@ RequiresMountsFor=%t/containers
...
+ExecStartPre=/bin/bash /usr/bin/microshift-containerized # not autogenerated
 ExecStartPre=/bin/rm -f %t/%n.ctr-id
-ExecStart=/usr/bin/podman run --cidfile=%t/%n.ctr-id --cgroups=no-conmon --rm --replace --sdnotify=container --label io.containers.autoupdate=registry --network=host --privileged -d --name microshift -v /var/run:/var/run -v /sys:/sys:ro -v /var/lib:/var/lib:rw,rshared -v /lib/modules:/lib/modules -v /etc:/etc -v /run/containers:/run/containers -v /var/log:/var/log -e KUBECONFIG=/var/lib/microshift/resources/kubeadmin/kubeconfig quay.io/microshift/microshift:latest
+ExecStart=/usr/bin/podman run --cidfile=%t/%n.ctr-id --cgroups=no-conmon --rm --replace --sdnotify=container --label io.containers.autoupdate=registry --network=host --privileged -d --name microshift -v /var/run/crio/crio.sock:/var/run/crio/crio.sock:rw,rshared -v /var/lib/microshift:/var/lib/microshift:rw,rshared -v /var/lib/kubelet:/var/lib/kubelet:rw,rshared -v /var/log:/var/log -e KUBECONFIG=/var/lib/microshift/resources/kubeadmin/kubeconfig quay.io/microshift/microshift:latest
```
Comment (mangelajo) [inline packaging/systemd/microshift-containerized.service:19]: This still has

`-v /var/lib/microshift:/var/lib/microshift:rw,rshared` 
right?

File: packaging/systemd/microshift-containerized-configure-selinux [hunk 3]
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
+
+set -euxo pipefail
+
+sudo dnf -y install container-selinux
+sudo dnf -y install selinux-policy-devel 
+curl -L -o /tmp/microshift.fc https://raw.githubusercontent.com/redhat-et/microshift/main/packaging/selinux/microshift.fc
+curl -L -o /tmp/microshift.te https://raw.githubusercontent.com/redhat-et/microshift/main/packaging/selinux/microshift.te
+make -f /usr/share/selinux/devel/Makefile -C /tmp
+sudo dnf -y remove selinux-policy-devel
```
Comment (mangelajo) [inline packaging/systemd/microshift-containerized-configure-selinux:10]:
```suggestion
```

If we keep this script, please let's stop removing the devel package :-)

install.sh removes it on my system, which I never expect and it's annoying.

File: packaging/systemd/microshift-containerized.service (hunk 1 above)
Comment (mangelajo) [reply packaging/systemd/microshift-containerized.service:16]: https://github.com/redhat-et/microshift/blob/main/packaging/rpm/microshift.spec#L226

**Review by mangelajo — CHANGES_REQUESTED** [review]
see comments.

File: packaging/systemd/microshift-containerized.service (hunk 2 above)
Comment (husky-parul) [reply packaging/systemd/microshift-containerized.service:19]: Yes. We are going ahead with `/var/lib/microshift:/var/lib/microshift:rw,rshared` instead of `microshift-data:/var/lib/microshift:rw,rshared`

**Comment by husky-parul** [comment]
@mangelajo Ready for review

File: packaging/systemd/microshift-containerized [hunk 4]
```diff
@@ -2,7 +2,5 @@
 
 set -euxo pipefail
 
-[[ -d /etc/microshift-containerized ]] || mkdir -p /etc/microshift-containerized
-cat <<EOF > /etc/microshift-containerized/microshift-containerized.conf
-export KUBECONFIG=/var/lib/microshift/resources/kubeadmin/kubeconfig
-EOF
+setsebool -P container_manage_cgroup true
+sudo mkdir -p /var/lib/microshift
```
Comment (fzdarsky) [inline packaging/systemd/microshift-containerized:6]: Can we fold this file completely into the `microshift-containerized.service`, such that the UX is like we document in https://microshift.io/docs/getting-started/ ?

File: packaging/systemd/microshift-containerized-configure-selinux [hunk 5]
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
```
Comment (fzdarsky) [inline packaging/systemd/microshift-containerized-configure-selinux:1]: Is this file still needed / up-to-date?

If we're running containerized, only the crio.sock should be bind-mounted from the host (plus the `/var/lib/microshift` datadir if we're using hostdir instead of voliumes). So why do we still need to create these other dirs and adjust SELinux policies?

Comment (husky-parul) [reply packaging/systemd/microshift-containerized-configure-selinux:1]: We don't need to adjust the selinux anymore. From @mangelajo comment. I will remove the files as well. It was based on the assumption that the users didn't download the rpms and would be only using the `microshift-containerized.service`
_"Let's assume that people will download the rpms for microshift via dnf, which will include both microshift-containerized + the selinux modules compiled."_

Comment (fzdarsky) [reply packaging/systemd/microshift-containerized-configure-selinux:1]: Wait... I think we're confusing deployment models here:
The containerised deployment is for users who want to deploy MIcroShift at runtime on top of a R4E container host. That is, they cannot dynamically install an .RPM (that would need to be baked into the rpm-ostree).

Comment (mangelajo) [reply packaging/systemd/microshift-containerized-configure-selinux:1]: Yes, they can't dynamically install it, but if you want to bake it, then we need it also to be in the RPM?.

In any case I don't think we should be running make for the selinux files on installation, or downloading seliux-policy-devel etc... in that case we should probably provide the selinux compiled policy, if that's standard across kernel versions, which I don't know.

May be the container itself should be built by installing the RPM we have?, then the user just downloads the .service file? Can we install an selinux policy from inside a privileged container?

Comment (mangelajo) [reply packaging/systemd/microshift-containerized-configure-selinux:1]: According to this, can’t just people install an rpm? We made a subpackage for the microshift container service which will also pull the selinux defs: https://developers.redhat.com/blog/2020/03/12/how-to-customize-fedora-coreos-for-dedicated-workloads-with-ostree

Comment (fzdarsky) [reply packaging/systemd/microshift-containerized-configure-selinux:1]: What I meant is we have two deployment models:
1) The .rpm-based installation, which is for embedding into rpm-ostrees and runs MicroShift as binary on the host / non-containerized. We have corresponding SELinux rules for that.
2) The containerized deployment, which is for users who _don't_ want to embed into an rpm-ostree but instead use a container host and add/remove MicroShift at runtime - without reboot. This is where it doesn't make sense to install the sytemd unit via an .rpm. Instead, it'll need to be created via configuration management resp. from the application management agent.

**Comment by husky-parul** [comment]
@fzdarsky consolidated everything in microshift-containerized.service
needs ok to test

File: packaging/systemd/microshift-containerized.service [hunk 6]
```diff
@@ -13,8 +13,10 @@ RequiresMountsFor=%t/containers
 Environment=PODMAN_SYSTEMD_UNIT=%n
 Restart=on-failure
 TimeoutStopSec=70
+ExecStartPre=/bin/mkdir -p /var/lib/microshift # not autogenerated
+ExecStartPre=setsebool container_manage_cgroup true # not autogenerated
 ExecStartPre=/bin/rm -f %t/%n.ctr-id
-ExecStart=/usr/bin/podman run --cidfile=%t/%n.ctr-id --cgroups=no-conmon --rm --replace --sdnotify=container --label io.containers.autoupdate=registry --network=host --privileged -d --name microshift -v /var/run:/var/run -v /sys:/sys:ro -v /var/lib:/var/lib:rw,rshared -v /lib/modules:/lib/modules -v /etc:/etc -v /run/containers:/run/containers -v /var/log:/var/log -e KUBECONFIG=/var/lib/microshift/resources/kubeadmin/kubeconfig quay.io/microshift/microshift:latest
+ExecStart=/usr/bin/podman run --cidfile=%t/%n.ctr-id --cgroups=no-conmon --rm --replace --sdnotify=container --label io.containers.autoupdate=registry --network=host --privileged -d --name microshift -v /var/run/crio/crio.sock:/var/run/crio/crio.sock:rw,rshared -v /var/lib/microshift:/var/lib/microshift:rw,rshared -v /var/lib/kubelet:/var/lib/kubelet:rw,rshared -v /var/log:/var/log quay.io/microshift/microshift:latest
```
Comment (fzdarsky) [inline packaging/systemd/microshift-containerized.service:19]: These Podman arguments still bind mount the `/var/lib/microshift` data-dir to the host file system instead of using a Podman volume `microshift-data` as described in #433.

Also, it's not clear why we need to bind mount `/var/lib/kubelet` and `/var/log` at all? For the latter, I would have expected the container to simply log to stdout as is common on K8s?

**Comment by cooktheryan** [comment]
@husky-parul I'm not sure if this directory creation should be included with the start up script but without making the directory for /var/lib/kubelet things fail 


```
Jan 10 18:47:39 fedora setsebool[15691]: The container_manage_cgroup policy boolean was changed to true by root
Jan 10 18:47:39 fedora bash[15690]: + /usr/bin/podman volume exists microshift-data
Jan 10 18:47:40 fedora podman[15745]: Error: statfs /var/lib/kubelet: no such file or directory
Jan 10 18:47:40 fedora systemd[1]: microshift.service: Main process exited, code=exited, status=125/n/a
Jan 10 18:47:40 fedora podman[15793]: Error: error reading CIDFile: open /run/microshift.service.ctr-id: no such file or directory
Jan 10 18:47:40 fedora systemd[1]: microshift.service: Control process exited, code=exited, status=125/n/a
Jan 10 18:47:40 fedora systemd[1]: microshift.service: Failed with result 'exit-code'.
Jan 10 18:47:40 fedora systemd[1]: Failed to start MicroShift Containerized # not autogenerated.
Jan 10 18:47:40 fedora systemd[1]: microshift.service: Scheduled restart job, restart counter is at 5.
Jan 10 18:47:40 fedora systemd[1]: Stopped MicroShift Containerized # not autogenerated.
Jan 10 18:47:40 fedora systemd[1]: microshift.service: Start request repeated too quickly.
Jan 10 18:47:40 fedora systemd[1]: microshift.service: Failed with result 'exit-code'.
Jan 10 18:47:40 fedora systemd[1]: Failed to start MicroShift Containerized # not autogenerated.
^C
[root@fedora ~]# mkdir /var/lib/kubelet
[root@fedora ~]# systemctl start microshift
[root@fedora ~]# journalctl -flu microshift

File: packaging/systemd/microshift-containerized.service [hunk 7]
```diff
@@ -13,9 +13,9 @@ RequiresMountsFor=%t/containers
 Environment=PODMAN_SYSTEMD_UNIT=%n
 Restart=on-failure
 TimeoutStopSec=70
+ExecStartPre=/bin/bash /usr/bin/microshift-containerized # not autogenerated
 ExecStartPre=/bin/rm -f %t/%n.ctr-id
-ExecStartPre=/usr/bin/mkdir -p /var/hpvolumes
-ExecStart=/usr/bin/podman run --cidfile=%t/%n.ctr-id --cgroups=no-conmon --rm --replace --sdnotify=container --label io.containers.autoupdate=registry --network=host --privileged -d --name microshift -v /var/run:/var/run -v /sys:/sys:ro -v /var/lib:/var/lib:rw,rshared -v /var/hpvolumes:/var/hpvolumes -v /lib/modules:/lib/modules -v /etc:/etc -v /run/containers:/run/containers -v /var/log:/var/log -e KUBECONFIG=/var/lib/microshift/resources/kubeadmin/kubeconfig quay.io/microshift/microshift:latest
+ExecStart=/usr/bin/podman run --cidfile=%t/%n.ctr-id --cgroups=no-conmon --rm --replace --sdnotify=container --label io.containers.autoupdate=registry --network=host --privileged -d --name microshift -v /var/run/crio/crio.sock:/var/run/crio/crio.sock:rw,rshared -v microshift-data:/var/lib/microshift:rw,rshared -v /var/lib/kubelet:/var/lib/kubelet:rw,rshared -v /var/log:/var/log quay.io/microshift/microshift:4.8.0-0.microshift-2022-01-04-205130
```
Comment (cooktheryan) [inline packaging/systemd/microshift-containerized.service:18]: Should we bump the version of the image to the latest 4.8 build or worst case latest?

**Comment by husky-parul** [comment]
> 

@cooktheryan From what I gathered from the community meeting, the systemd file is not supposed to do anything except for starting and managing the lifecycle so I am moving the `mkdir -p /var/libe/kubelet` to docs.

**Comment by fzdarsky** [comment]
> > 
> 
> @cooktheryan From what I gathered from the community meeting, the systemd file is not supposed to do anything except for starting and managing the lifecycle so I am moving the `mkdir -p /var/libe/kubelet` to docs.

@husky-parul I think having this in the `ExecStartPre` instead of the doc is fine as it's part of the pre-reqs for starting.
Anyway, let's get this merged and possibly add in a follow-up?

/lgtm

**Comment by cooktheryan** [comment]
Woot woot!!!!!
